File: packages/kits/NeuronKit/Sources/NeuronKit/Reduction/QueryDateWindow.swift
import Foundation

// QueryDateWindow — deterministic absolute-date-expression parsing for the
// temporal_recall recipe (decision record
// DECISION_SEARCH_STRATEGY_RECIPES_2026-08-19, item 4).
//
// The recall lanes measure content similarity and currency; none of them
// reads a date stated in the QUERY ("in July", "on 8 May 2023") and matches
// it against drawer event_time. This module supplies that reading: it parses
// explicit absolute date expressions out of a query string and returns the
// UTC windows they name. Version 1 is absolute-only by design ruling —
// relative expressions ("last week") need a reference clock and are a
// separate, later surface.
//
// Determinism: pure string→window computation. No Date(), no Calendar
// locale, no current time anywhere — the same query yields the same windows
// on every machine forever. Month-only expressions cannot know a year, so
// they return one window PER YEAR in a caller-supplied plausible range; the
// recipe intersects those with the estate's actual event-time span.

/// One inclusive UTC window an absolute date expression names.
public struct QueryDateWindow: Sendable, Equatable {
    /// Inclusive window start, "YYYY-MM-DDT00:00:00Z" form.
    public let start: String
    /// Inclusive window end, "YYYY-MM-DDT23:59:59Z" form.
    public let end: String
    /// The query substring the window was parsed from (diagnostic).
    public let matchedText: String

    public init(start: String, end: String, matchedText: String) {
        self.start = start
        self.end = end
        self.matchedText = matchedText
    }
}

/// The result of parsing a query for absolute date expressions.
public enum QueryDateParse: Sendable, Equatable {
    /// No absolute date expression found.
    case none
    /// Fully-anchored windows (year known).
    case anchored([QueryDateWindow])
    /// Month-only expression: the month is known, the year is not.
    /// The recipe expands this against the estate's own year span.
    case monthOnly(month: Int, matchedText: String)
}

private let queryMonthNames: [String: Int] = [
    "january": 1, "february": 2, "march": 3, "april": 4, "may": 5,
    "june": 6, "july": 7, "august": 8, "september": 9, "october": 10,
    "november": 11, "december": 12,
]

/// Day counts per month; February resolved with the standard leap rule.
private func daysIn(month: Int, year: Int) -> Int {
    switch month {
    case 1, 3, 5, 7, 8, 10, 12: return 31
    case 4, 6, 9, 11: return 30
    default:
        let leap = (year % 4 == 0 && year % 100 != 0) || year % 400 == 0
        return leap ? 29 : 28
    }
}

private func dayWindow(_ y: Int, _ m: Int, _ d: Int, _ text: String) -> QueryDateWindow {
    QueryDateWindow(
        start: String(format: "%04d-%02d-%02dT00:00:00Z", y, m, d),
        end: String(format: "%04d-%02d-%02dT23:59:59Z", y, m, d),
        matchedText: text)
}

private func monthWindow(_ y: Int, _ m: Int, _ text: String) -> QueryDateWindow {
    QueryDateWindow(
        start: String(format: "%04d-%02d-01T00:00:00Z", y, m),
        end: String(format: "%04d-%02d-%02dT23:59:59Z", y, m, daysIn(month: m, year: y)),
        matchedText: text)
}

private func yearWindow(_ y: Int, _ text: String) -> QueryDateWindow {
    QueryDateWindow(
        start: String(format: "%04d-01-01T00:00:00Z", y),
        end: String(format: "%04d-12-31T23:59:59Z", y),
        matchedText: text)
}

/// Expands a month-only parse against a year range (typically the estate's
/// own event-time span): one month window per year, in ascending year order.
public func expandMonthOnly(
    month: Int, matchedText: String, years: ClosedRange<Int>
) -> [QueryDateWindow] {
    years.map { monthWindow($0, month, matchedText) }
}

/// Parses the first absolute date expression in a query.
///
/// Grammar v1 (design ruling Q2 — absolute only):
///   day-month-year   "8 May 2023" / "8 May, 2023" / "May 8, 2023" → day window
///   ISO date         "2023-05-08"                                  → day window
///   month-year       "May 2023" / "in May 2023"                    → month window
///   month-only       "in July" / bare month name                   → monthOnly
///   year-only        "in 2023" / bare 1900–2099 number             → year window
///
/// Longest-anchor-wins: a day-month-year match is preferred over the
/// month-year reading of the same tokens, month-year over month-only.
/// Month names inside other words do not match (word-boundary tokenizing).
public func parseQueryDateExpression(_ query: String) -> QueryDateParse {
    // Word-boundary tokens, lowercased, punctuation stripped at the edges.
    let rawTokens = query.split(whereSeparator: { $0 == " " || $0 == "\n" || $0 == "\t" })
    let tokens: [String] = rawTokens.map {
        $0.lowercased().trimmingCharacters(in: CharacterSet(charactersIn: ".,!?;:()'\""))
    }

    func int(_ s: String) -> Int? { Int(s) }
    func isYear(_ n: Int) -> Bool { (1900...2099).contains(n) }
    func isDay(_ n: Int) -> Bool { (1...31).contains(n) }

    // ISO date token anywhere: YYYY-MM-DD.
    for t in tokens where t.count == 10 && t[t.index(t.startIndex, offsetBy: 4)] == "-" {
        let p = t.split(separator: "-").map(String.init)
        if p.count == 3, let y = Int(p[0]), let m = Int(p[1]), let d = Int(p[2]),
           isYear(y), (1...12).contains(m), isDay(d) {
            return .anchored([dayWindow(y, m, d, t)])
        }
    }

    for (i, t) in tokens.enumerated() {
        guard let month = queryMonthNames[t] else { continue }
        let prev = i > 0 ? tokens[i - 1] : ""
        let next = i + 1 < tokens.count ? tokens[i + 1] : ""
        let next2 = i + 2 < tokens.count ? tokens[i + 2] : ""
        // "8 May 2023" / "8 May, 2023"
        if let d = int(prev), isDay(d), let y = int(next), isYear(y) {
            return .anchored([dayWindow(y, month, d, "\(prev) \(t) \(next)")])
        }
        // "May 8, 2023"
        if let d = int(next), isDay(d), let y = int(next2), isYear(y) {
            return .anchored([dayWindow(y, month, d, "\(t) \(next) \(next2)")])
        }
        // "8 May" with no year: month-only is the honest reading (the year is
        // unstated); day precision without a year would guess.
        // "May 2023"
        if let y = int(next), isYear(y) {
            return .anchored([monthWindow(y, month, "\(t) \(next)")])
        }
        // bare month → month-only
        return .monthOnly(month: month, matchedText: t)
    }

    // year-only: "in 2023" or a bare plausible year token
    for t in tokens {
        if let y = int(t), isYear(y) {
            return .anchored([yearWindow(y, String(y))])
        }
    }
    return .none
}

/// True when an ISO event-time string falls inside the window (inclusive).
/// String comparison is correct because all values share the fixed-width
/// "YYYY-MM-DDTHH:MM:SSZ" shape (ISO8601 UTC sorts lexicographically).
public func windowContains(_ window: QueryDateWindow, eventTime: String) -> Bool {
    eventTime >= window.start && eventTime <= window.end
}

/// Returns the window widened by `days` on each side (the sliding-window
/// expansion, ruling Q2 2026-08-19: expand ±1 day at a time until enough
/// candidates exist to compare, hard cap ±10). Pure civil-calendar
/// arithmetic on the fixed "YYYY-MM-DD…" bound shape; the time-of-day
/// parts of each bound are preserved (start keeps 00:00:00, end keeps
/// 23:59:59). `days <= 0` returns the window unchanged.
public func paddedWindow(_ window: QueryDateWindow, days: Int) -> QueryDateWindow {
    guard days > 0 else { return window }
    return QueryDateWindow(
        start: shiftISODay(window.start, by: -days),
        end: shiftISODay(window.end, by: days),
        matchedText: window.matchedText)
}

/// Shifts the DATE part of a fixed-shape ISO string by `days`, preserving
/// the time-of-day suffix. Hinnant civil-from-days / days-from-civil pure
/// integer math — no Calendar, no clock, identical in the Rust twin.
func shiftISODay(_ iso: String, by days: Int) -> String {
    let y = Int(iso.prefix(4))!
    let m = Int(iso.dropFirst(5).prefix(2))!
    let d = Int(iso.dropFirst(8).prefix(2))!
    let suffix = String(iso.dropFirst(10))
    // days_from_civil (Hinnant)
    let yy = m <= 2 ? y - 1 : y
    let era = (yy >= 0 ? yy : yy - 399) / 400
    let yoe = yy - era * 400
    let mp = (m + 9) % 12
    let doy = (153 * mp + 2) / 5 + d - 1
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy
    var z = era * 146_097 + doe - 719_468
    z += days
    // civil_from_days (Hinnant)
    let z2 = z + 719_468
    let era2 = (z2 >= 0 ? z2 : z2 - 146_096) / 146_097
    let doe2 = z2 - era2 * 146_097
    let yoe2 = (doe2 - doe2 / 1460 + doe2 / 36_524 - doe2 / 146_096) / 365
    let yr = yoe2 + era2 * 400
    let doy2 = doe2 - (365 * yoe2 + yoe2 / 4 - yoe2 / 100)
    let mp2 = (5 * doy2 + 2) / 153
    let day = doy2 - (153 * mp2 + 2) / 5 + 1
    let mon = mp2 < 10 ? mp2 + 3 : mp2 - 9
    let year = mon <= 2 ? yr + 1 : yr
    return String(format: "%04d-%02d-%02d", year, mon, day) + suffix
}

/// True when the query ASKS FOR a date rather than stating one ("When did
/// Melanie go camping?", "What date was the gala?") — the Gap 3 scanner
/// fork (DECISION_DENSE_LANE_ENRICHMENT program). Deterministic token-
/// bigram scan; a query can be date-seeking AND carry an absolute date
/// ("When in 2023 did…"), in which case the parsed window wins upstream.
public func isDateSeekingQuery(_ query: String) -> Bool {
    let tokens = query.lowercased().split(separator: " ").map {
        $0.trimmingCharacters(in: CharacterSet(charactersIn: ".,!?;:()'\""))
    }
    // "how long ago" — the one trigram form.
    for i in 0..<max(tokens.count, 2) - 2 where tokens.count >= 3 {
        if tokens[i] == "how" && tokens[i + 1] == "long" && tokens[i + 2] == "ago" {
            return true
        }
    }
    guard tokens.count >= 2 else { return false }
    for i in 0..<(tokens.count - 1) {
        let a = tokens[i], b = tokens[i + 1]
        if a == "when" && (b == "did" || b == "was" || b == "will" || b == "is") {
            return true
        }
        if (a == "what" || a == "which") && (b == "date" || b == "day" || b == "year" || b == "month") {
            return true
        }
    }
    return false
}

File: packages/kits/NeuronKit/rust/src/query_date_window.rs
//! QueryDateWindow — deterministic absolute-date-expression parsing for the
//! temporal_recall recipe. Twin of Swift
//! `NeuronKit/Reduction/QueryDateWindow.swift`; see that file for the design
//! rationale (decision record DECISION_SEARCH_STRATEGY_RECIPES_2026-08-19,
//! item 4). Pure string→window computation: no clock, no locale.

/// One inclusive UTC window an absolute date expression names.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct QueryDateWindow {
    /// Inclusive window start, "YYYY-MM-DDT00:00:00Z" form.
    pub start: String,
    /// Inclusive window end, "YYYY-MM-DDT23:59:59Z" form.
    pub end: String,
    /// The query substring the window was parsed from (diagnostic).
    pub matched_text: String,
}

/// The result of parsing a query for absolute date expressions.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum QueryDateParse {
    /// No absolute date expression found.
    None,
    /// Fully-anchored windows (year known).
    Anchored(Vec<QueryDateWindow>),
    /// Month-only expression: month known, year not — the recipe expands
    /// this against the estate's own year span.
    MonthOnly { month: u32, matched_text: String },
}

fn month_number(name: &str) -> Option<u32> {
    match name {
        "january" => Some(1), "february" => Some(2), "march" => Some(3),
        "april" => Some(4), "may" => Some(5), "june" => Some(6),
        "july" => Some(7), "august" => Some(8), "september" => Some(9),
        "october" => Some(10), "november" => Some(11), "december" => Some(12),
        _ => None,
    }
}

/// Day counts per month; February resolved with the standard leap rule.
fn days_in(month: u32, year: i32) -> u32 {
    match month {
        1 | 3 | 5 | 7 | 8 | 10 | 12 => 31,
        4 | 6 | 9 | 11 => 30,
        _ => {
            let leap = (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
            if leap { 29 } else { 28 }
        }
    }
}

fn day_window(y: i32, m: u32, d: u32, text: &str) -> QueryDateWindow {
    QueryDateWindow {
        start: format!("{y:04}-{m:02}-{d:02}T00:00:00Z"),
        end: format!("{y:04}-{m:02}-{d:02}T23:59:59Z"),
        matched_text: text.to_string(),
    }
}

fn month_window(y: i32, m: u32, text: &str) -> QueryDateWindow {
    QueryDateWindow {
        start: format!("{y:04}-{m:02}-01T00:00:00Z"),
        end: format!("{y:04}-{m:02}-{:02}T23:59:59Z", days_in(m, y)),
        matched_text: text.to_string(),
    }
}

fn year_window(y: i32, text: &str) -> QueryDateWindow {
    QueryDateWindow {
        start: format!("{y:04}-01-01T00:00:00Z"),
        end: format!("{y:04}-12-31T23:59:59Z"),
        matched_text: text.to_string(),
    }
}

/// Expands a month-only parse against a year range: one month window per
/// year, ascending. Twin of Swift `expandMonthOnly`.
pub fn expand_month_only(
    month: u32, matched_text: &str, years: std::ops::RangeInclusive<i32>,
) -> Vec<QueryDateWindow> {
    years.map(|y| month_window(y, month, matched_text)).collect()
}

/// Parses the first absolute date expression in a query. Grammar and
/// precedence are the Swift twin's, byte for byte on the golden pins.
pub fn parse_query_date_expression(query: &str) -> QueryDateParse {
    let tokens: Vec<String> = query
        .split_whitespace()
        .map(|t| t.to_lowercase().trim_matches(|c: char| ".,!?;:()'\"".contains(c)).to_string())
        .collect();

    let is_year = |n: i32| (1900..=2099).contains(&n);
    let is_day = |n: i32| (1..=31).contains(&n);

    // ISO date token anywhere: YYYY-MM-DD.
    for t in &tokens {
        if t.len() == 10 && t.as_bytes().get(4) == Some(&b'-') {
            let p: Vec<&str> = t.split('-').collect();
            if p.len() == 3 {
                if let (Ok(y), Ok(m), Ok(d)) =
                    (p[0].parse::<i32>(), p[1].parse::<u32>(), p[2].parse::<u32>())
                {
                    if is_year(y) && (1..=12).contains(&m) && is_day(d as i32) {
                        return QueryDateParse::Anchored(vec![day_window(y, m, d, t)]);
                    }
                }
            }
        }
    }

    for (i, t) in tokens.iter().enumerate() {
        let Some(month) = month_number(t) else { continue };
        let prev = if i > 0 { tokens[i - 1].as_str() } else { "" };
        let next = tokens.get(i + 1).map(String::as_str).unwrap_or("");
        let next2 = tokens.get(i + 2).map(String::as_str).unwrap_or("");
        // "8 May 2023"
        if let (Ok(d), Ok(y)) = (prev.parse::<i32>(), next.parse::<i32>()) {
            if is_day(d) && is_year(y) {
                return QueryDateParse::Anchored(vec![day_window(
                    y, month, d as u32, &format!("{prev} {t} {next}"))]);
            }
        }
        // "May 8, 2023"
        if let (Ok(d), Ok(y)) = (next.parse::<i32>(), next2.parse::<i32>()) {
            if is_day(d) && is_year(y) {
                return QueryDateParse::Anchored(vec![day_window(
                    y, month, d as u32, &format!("{t} {next} {next2}"))]);
            }
        }
        // "May 2023"
        if let Ok(y) = next.parse::<i32>() {
            if is_year(y) {
                return QueryDateParse::Anchored(vec![month_window(
                    y, month, &format!("{t} {next}"))]);
            }
        }
        // bare month → month-only (a day without a year also lands here;
        // day precision without a year would guess).
        return QueryDateParse::MonthOnly { month, matched_text: t.clone() };
    }

    // year-only
    for t in &tokens {
        if let Ok(y) = t.parse::<i32>() {
            if is_year(y) {
                return QueryDateParse::Anchored(vec![year_window(y, &y.to_string())]);
            }
        }
    }
    QueryDateParse::None
}

/// Returns the window widened by `days` on each side (the sliding-window
/// expansion, ruling Q2 2026-08-19: expand ±1 day at a time until enough
/// candidates exist to compare, hard cap ±10). Pure civil-calendar
/// arithmetic; time-of-day suffixes are preserved. `days <= 0` returns the
/// window unchanged. Twin of Swift `paddedWindow`.
pub fn padded_window(window: &QueryDateWindow, days: i64) -> QueryDateWindow {
    if days <= 0 {
        return window.clone();
    }
    QueryDateWindow {
        start: shift_iso_day(&window.start, -days),
        end: shift_iso_day(&window.end, days),
        matched_text: window.matched_text.clone(),
    }
}

/// Shifts the DATE part of a fixed-shape ISO string by `days`, preserving
/// the time-of-day suffix. Hinnant days-from-civil / civil-from-days pure
/// integer math — no clock; twin of Swift `shiftISODay`.
pub fn shift_iso_day(iso: &str, days: i64) -> String {
    let y: i64 = iso[0..4].parse().unwrap_or(1970);
    let m: i64 = iso[5..7].parse().unwrap_or(1);
    let d: i64 = iso[8..10].parse().unwrap_or(1);
    let suffix = &iso[10..];
    // days_from_civil (Hinnant)
    let yy = if m <= 2 { y - 1 } else { y };
    let era = if yy >= 0 { yy } else { yy - 399 } / 400;
    let yoe = yy - era * 400;
    let mp = (m + 9) % 12;
    let doy = (153 * mp + 2) / 5 + d - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    let z = era * 146_097 + doe - 719_468 + days;
    // civil_from_days (Hinnant)
    let z2 = z + 719_468;
    let era2 = if z2 >= 0 { z2 } else { z2 - 146_096 } / 146_097;
    let doe2 = z2 - era2 * 146_097;
    let yoe2 = (doe2 - doe2 / 1460 + doe2 / 36_524 - doe2 / 146_096) / 365;
    let yr = yoe2 + era2 * 400;
    let doy2 = doe2 - (365 * yoe2 + yoe2 / 4 - yoe2 / 100);
    let mp2 = (5 * doy2 + 2) / 153;
    let day = doy2 - (153 * mp2 + 2) / 5 + 1;
    let mon = if mp2 < 10 { mp2 + 3 } else { mp2 - 9 };
    let year = if mon <= 2 { yr + 1 } else { yr };
    format!("{year:04}-{mon:02}-{day:02}{suffix}")
}

/// True when an ISO event-time string falls inside the window (inclusive).
/// Lexicographic comparison is correct for the fixed-width UTC ISO shape.
pub fn window_contains(window: &QueryDateWindow, event_time: &str) -> bool {
    event_time >= window.start.as_str() && event_time <= window.end.as_str()
}

/// True when the query ASKS FOR a date rather than stating one. Twin of
/// Swift `isDateSeekingQuery` (Gap 3 scanner fork). Deterministic
/// token-bigram scan.
pub fn is_date_seeking_query(query: &str) -> bool {
    let tokens: Vec<String> = query
        .split_whitespace()
        .map(|t| t.to_lowercase().trim_matches(|c: char| ".,!?;:()'\"".contains(c)).to_string())
        .collect();
    if tokens.len() >= 3 {
        for w in tokens.windows(3) {
            if w[0] == "how" && w[1] == "long" && w[2] == "ago" {
                return true;
            }
        }
    }
    for w in tokens.windows(2) {
        let (a, b) = (w[0].as_str(), w[1].as_str());
        if a == "when" && matches!(b, "did" | "was" | "will" | "is") {
            return true;
        }
        if matches!(a, "what" | "which") && matches!(b, "date" | "day" | "year" | "month") {
            return true;
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;

    // Golden pins — literal twins of QueryDateWindowTests.swift.
    #[test]
    fn day_month_year_pins() {
        assert_eq!(
            parse_query_date_expression("What happened on 8 May 2023 at the fair?"),
            QueryDateParse::Anchored(vec![QueryDateWindow {
                start: "2023-05-08T00:00:00Z".into(),
                end: "2023-05-08T23:59:59Z".into(),
                matched_text: "8 may 2023".into(),
            }]));
        assert_eq!(
            parse_query_date_expression("The 2023-05-08 entry"),
            QueryDateParse::Anchored(vec![QueryDateWindow {
                start: "2023-05-08T00:00:00Z".into(),
                end: "2023-05-08T23:59:59Z".into(),
                matched_text: "2023-05-08".into(),
            }]));
    }

    #[test]
    fn month_year_leap_pin() {
        assert_eq!(
            parse_query_date_expression("back in February 2024"),
            QueryDateParse::Anchored(vec![QueryDateWindow {
                start: "2024-02-01T00:00:00Z".into(),
                end: "2024-02-29T23:59:59Z".into(),
                matched_text: "february 2024".into(),
            }]));
    }

    #[test]
    fn month_only_pin() {
        assert_eq!(
            parse_query_date_expression("When did Melanie go camping in July?"),
            QueryDateParse::MonthOnly { month: 7, matched_text: "july".into() });
        assert_eq!(
            expand_month_only(7, "july", 2022..=2023),
            vec![
                QueryDateWindow { start: "2022-07-01T00:00:00Z".into(),
                                  end: "2022-07-31T23:59:59Z".into(),
                                  matched_text: "july".into() },
                QueryDateWindow { start: "2023-07-01T00:00:00Z".into(),
                                  end: "2023-07-31T23:59:59Z".into(),
                                  matched_text: "july".into() },
            ]);
    }

    #[test]
    fn year_only_and_none() {
        assert_eq!(
            parse_query_date_expression("everything from 2023"),
            QueryDateParse::Anchored(vec![QueryDateWindow {
                start: "2023-01-01T00:00:00Z".into(),
                end: "2023-12-31T23:59:59Z".into(),
                matched_text: "2023".into(),
            }]));
        assert_eq!(parse_query_date_expression("What is Melanie's favorite song?"),
                   QueryDateParse::None);
        assert_eq!(parse_query_date_expression("mayonnaise recipes and marching bands"),
                   QueryDateParse::None);
    }

    #[test]
    fn day_without_year_is_month_only() {
        assert_eq!(
            parse_query_date_expression("they met on 8 May at the market"),
            QueryDateParse::MonthOnly { month: 5, matched_text: "may".into() });
    }

    #[test]
    fn date_seeking_pins() {
        // Literal twins of the Swift isDateSeekingQuery pins.
        assert!(is_date_seeking_query("When did Melanie go camping?"));
        assert!(is_date_seeking_query("What date was the gala in Boston?"));
        assert!(is_date_seeking_query("how long ago did they meet"));
        assert!(!is_date_seeking_query("Which city was Calvin at on October 3, 2023?"));
        assert!(!is_date_seeking_query("What is Melanie's favorite song?"));
    }

    #[test]
    fn padded_window_pins() {
        // Literal twins of Swift paddedWindowPins.
        let w = QueryDateWindow { start: "2023-10-03T00:00:00Z".into(),
                                  end: "2023-10-03T23:59:59Z".into(),
                                  matched_text: "x".into() };
        let p1 = padded_window(&w, 1);
        assert_eq!(p1.start, "2023-10-02T00:00:00Z");
        assert_eq!(p1.end, "2023-10-04T23:59:59Z");
        let feb = QueryDateWindow { start: "2024-03-01T00:00:00Z".into(),
                                    end: "2024-03-01T23:59:59Z".into(),
                                    matched_text: "x".into() };
        let p2 = padded_window(&feb, 1);
        assert_eq!(p2.start, "2024-02-29T00:00:00Z");
        assert_eq!(p2.end, "2024-03-02T23:59:59Z");
        let jan = QueryDateWindow { start: "2023-01-05T00:00:00Z".into(),
                                    end: "2023-01-05T23:59:59Z".into(),
                                    matched_text: "x".into() };
        let p3 = padded_window(&jan, 10);
        assert_eq!(p3.start, "2022-12-26T00:00:00Z");
        assert_eq!(p3.end, "2023-01-15T23:59:59Z");
        assert_eq!(padded_window(&w, 0), w);
    }

    #[test]
    fn containment_inclusive() {
        let w = QueryDateWindow { start: "2023-07-01T00:00:00Z".into(),
                                  end: "2023-07-31T23:59:59Z".into(),
                                  matched_text: "july".into() };
        assert!(window_contains(&w, "2023-07-17T14:31:00Z"));
        assert!(window_contains(&w, "2023-07-01T00:00:00Z"));
        assert!(!window_contains(&w, "2023-08-01T00:00:00Z"));
    }
}
